File: src/node/src/group/group.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "group.h"

#include <cassert>
#include <cstdint>
#include <cstdlib>

#include "serializeddata.h"

namespace olive
{

#define super Node

NodeGroup::NodeGroup()
	: output_passthrough_(nullptr)
{
	set_flag(k_dont_show_in_create_menu);
}

std::string NodeGroup::name() const
{
	return "Group";
}

std::string NodeGroup::id() const
{
	return "org.olivevideoeditor.Olive.group";
}

std::vector<Node::CategoryID> NodeGroup::category() const
{
	return { k_category_unknown };
}

std::string NodeGroup::description() const
{
	return "A group of nodes that is represented as a single node.";
}

void NodeGroup::retranslate()
{
	super::retranslate();

	for (auto it = get_context_positions().cbegin();
		 it != get_context_positions().cend(); it++) {
		it->first->retranslate();
	}
}

bool NodeGroup::load_custom(XmlStreamReader *reader, SerializedData *data)
{
	while (xml_read_next_start_element(reader)) {
		if (reader->name() == "inputpassthroughs") {
			while (xml_read_next_start_element(reader)) {
				if (reader->name() == "inputpassthrough") {
					SerializedData::GroupLink link;

					link.group = this;

					while (xml_read_next_start_element(reader)) {
						if (reader->name() == "node") {
							link.input_node = strtoull(
								reader->read_element_text().c_str(), nullptr, 10);
						} else if (reader->name() == "input") {
							link.input_id = reader->read_element_text();
						} else if (reader->name() == "element") {
							link.input_element = int(strtol(
								reader->read_element_text().c_str(), nullptr, 10));
						} else if (reader->name() == "id") {
							link.passthrough_id = reader->read_element_text();
						} else if (reader->name() == "name") {
							link.custom_name = reader->read_element_text();
						} else if (reader->name() == "flags") {
							link.custom_flags = InputFlags(strtoull(
								reader->read_element_text().c_str(), nullptr, 10));
						} else if (reader->name() == "type") {
							link.data_type = NodeValue::get_data_type_from_name(
								reader->read_element_text());
						} else if (reader->name() == "default") {
							link.default_val = NodeValue::string_to_value(
								link.data_type, reader->read_element_text(),
								false);
						} else if (reader->name() == "properties") {
							while (xml_read_next_start_element(reader)) {
								if (reader->name() == "property") {
									std::string key;
									std::string value;

									while (xml_read_next_start_element(reader)) {
										if (reader->name() == "key") {
											key = reader->read_element_text();
										} else if (reader->name() == "value") {
											value = reader->read_element_text();
										} else {
											reader->skip_current_element();
										}
									}

									if (!key.empty()) {
										link.custom_properties.insert({ key,
																		value });
									}
								} else {
									reader->skip_current_element();
								}
							}
						} else {
							reader->skip_current_element();
						}
					}

					data->group_input_links.push_back(link);
				} else {
					reader->skip_current_element();
				}
			}
		} else if (reader->name() == "outputpassthrough") {
			data->group_output_links.insert(
				{ this, strtoull(reader->read_element_text().c_str(), nullptr,
								 10) });
		} else {
			reader->skip_current_element();
		}
	}

	return true;
}

void NodeGroup::save_custom(XmlStreamWriter *writer) const
{
	writer->write_start_element("inputpassthroughs");

	for (const NodeGroup::InputPassthrough &ip :
		 this->get_input_passthroughs()) {
		writer->write_start_element("inputpassthrough");

		// Reference to inner input
		writer->write_text_element(
			"node",
			std::to_string(reinterpret_cast<uintptr_t>(ip.second.node())));
		writer->write_text_element("input", ip.second.input());
		writer->write_text_element("element",
								 std::to_string(ip.second.element()));

		// ID of passthrough
		writer->write_text_element("id", ip.first);

		// Passthrough-specific details
		const std::string &input = ip.first;
		writer->write_text_element("name", this->Node::get_input_name(input));

		writer->write_text_element(
			"flags",
			std::to_string(
				(get_input_flags(input) & ~ip.second.get_flags()).value()));

		NodeValue::Type data_type = get_input_data_type(input);
		writer->write_text_element("type",
								 NodeValue::get_data_type_name(data_type));

		writer->write_text_element(
			"default",
			NodeValue::value_to_string(data_type, get_default_value(input), false));

		writer->write_start_element("properties");
		auto p = get_input_properties(input);
		for (auto it = p.cbegin(); it != p.cend(); it++) {
			writer->write_start_element("property");
			writer->write_text_element("key", it->first);
			writer->write_text_element("value", it->second.to_string());
			writer->write_end_element(); // property
		}
		writer->write_end_element(); // properties

		writer->write_end_element(); // input
	}

	writer->write_end_element(); // inputpassthroughs

	writer->write_text_element(
		"outputpassthrough",
		std::to_string(
			reinterpret_cast<uintptr_t>(this->get_output_passthrough())));
}

void NodeGroup::PostLoadEvent(SerializedData *data)
{
	super::PostLoadEvent(data);

	for (const SerializedData::GroupLink &l : data->group_input_links) {
		auto node_it = data->node_ptrs.find(l.input_node);
		if (node_it != data->node_ptrs.end()) {
			Node *input_node = node_it->second;
			NodeInput resolved(input_node, l.input_id, l.input_element);

			l.group->add_input_passthrough(resolved, l.passthrough_id);

			l.group->set_input_flag(l.passthrough_id,
								  InputFlag(l.custom_flags.value()));

			if (!l.custom_name.empty()) {
				l.group->set_input_name(l.passthrough_id, l.custom_name);
			}

			l.group->set_input_data_type(l.passthrough_id, l.data_type);

			l.group->set_default_value(l.passthrough_id, l.default_val);

			for (auto it = l.custom_properties.cbegin();
				 it != l.custom_properties.cend(); it++) {
				l.group->set_input_property(l.passthrough_id, it->first,
										  it->second);
			}
		}
	}

	for (auto it = data->group_output_links.cbegin();
		 it != data->group_output_links.cend(); it++) {
		auto node_it = data->node_ptrs.find(it->second);
		if (node_it != data->node_ptrs.end()) {
			it->first->set_output_passthrough(node_it->second);
		}
	}
}

std::string NodeGroup::add_input_passthrough(const NodeInput &input,
									   const std::string &force_id)
{
	assert(context_contains_node(input.node()));

	for (auto it = input_passthroughs_.cbegin();
		 it != input_passthroughs_.cend(); it++) {
		if (it->second == input) {
			// Already passing this input through
			return it->first;
		}
	}

	// Add input
	std::string id;
	if (force_id.empty()) {
		id = input.input();
		int i = 2;
		while (has_input_with_id(id)) {
			id = input.input() + "_" + std::to_string(i);
			i++;
		}
	} else {
		id = force_id;

		bool already_exists = false;
		for (auto it = input_passthroughs_.cbegin();
			 it != input_passthroughs_.cend(); it++) {
			if (it->first == id) {
				already_exists = true;
				break;
			}
		}

		assert(!already_exists);
	}

	add_input(id, input.get_data_type(), input.get_default_value(),
			 input.get_flags());

	input_passthroughs_.push_back({ id, input });

	return id;
}

void NodeGroup::remove_input_passthrough(const NodeInput &input)
{
	for (auto it = input_passthroughs_.begin(); it != input_passthroughs_.end();
		 it++) {
		if (it->second == input) {
			remove_input(it->first);
			input_passthroughs_.erase(it);
			break;
		}
	}
}

void NodeGroup::set_output_passthrough(Node *node)
{
	assert(!node || context_contains_node(node));

	output_passthrough_ = node;
}

bool NodeGroup::contains_input_passthrough(const NodeInput &input) const
{
	for (auto it = input_passthroughs_.cbegin();
		 it != input_passthroughs_.cend(); it++) {
		if (it->second == input) {
			return true;
		}
	}

	return false;
}

std::string NodeGroup::get_input_name(const std::string &id) const
{
	// If an override name was set, use that
	std::string override = super::get_input_name(id);
	if (!override.empty()) {
		return override;
	}

	// Call GetInputName of passed through node, which may be another group
	NodeInput pass = get_input_from_id(id);
	if (!pass.is_valid()) {
		return std::string();
	}
	return pass.node()->get_input_name(pass.input());
}

NodeInput NodeGroup::resolve_input(NodeInput input)
{
	while (get_inner(&input)) {
	}

	return input;
}

bool NodeGroup::get_inner(NodeInput *input)
{
	if (NodeGroup *g = dynamic_cast<NodeGroup *>(input->node())) {
		const NodeInput &passthrough = g->get_input_from_id(input->input());
		if (!passthrough.is_valid()) {
			return false;
		}

		input->set_node(passthrough.node());
		input->set_input(passthrough.input());
		return true;
	} else {
		return false;
	}
}

void NodeGroupAddInputPassthrough::redo()
{
	if (!group_->contains_input_passthrough(input_)) {
		group_->add_input_passthrough(input_, force_id_);
		actually_added_ = true;
	} else {
		actually_added_ = false;
	}
}

void NodeGroupAddInputPassthrough::undo()
{
	if (actually_added_) {
		group_->remove_input_passthrough(input_);
	}
}

void NodeGroupSetOutputPassthrough::redo()
{
	old_output_ = group_->get_output_passthrough();
	group_->set_output_passthrough(new_output_);
}

void NodeGroupSetOutputPassthrough::undo()
{
	group_->set_output_passthrough(old_output_);
}

}

File: src/node/rust/src/nodes/group.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Node group (C++ `src/node/src/group/group.{h,cpp}`,
//! `olive::NodeGroup`).
//!
//! A group wraps an inner node context and exposes selected inner
//! inputs as its own ("passthroughs"). The C++ class also declares the
//! undo commands `NodeGroupAddInputPassthrough` /
//! `NodeGroupSetOutputPassthrough`; those belong to the undo system
//! and are not modeled here.

use crate::factory::NodeMeta;
use crate::id::NodeId;
use crate::node::{Category, NodeBehavior, NodeCore};

/// Reference to an input of a node inside the group's context
/// (C++ `NodeInput`: node pointer + input id + array element).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct InnerInput {
	/// The inner node this input belongs to.
	pub node: NodeId,
	/// Input id on the inner node.
	pub input: String,
	/// Array element index (-1/0 for non-array inputs).
	pub element: i32,
}

/// One passthrough entry (C++ `NodeGroup::InputPassthrough`):
/// the group's own input id mapped to the inner input it mirrors.
pub type InputPassthrough = (String, InnerInput);

/// Node group node. Exposes inner-node inputs as its own inputs and
/// forwards its output from a designated inner node.
///
/// The C++ `output_passthrough_` is a raw `Node*`; here it is the
/// inner node's [`NodeId`] (`None` = unset). Note: the brief for this
/// port mentioned `connected_render_output`, but the actual C++
/// `NodeGroup` does not override it — only the methods below are
/// overridden.
pub struct NodeGroup {
	/// Passthrough table: group input id -> inner input
	/// (C++ `input_passthroughs_`).
	input_passthroughs: Vec<InputPassthrough>,
	/// Inner node whose output this group forwards
	/// (C++ `output_passthrough_`).
	output_passthrough: Option<NodeId>,
}

impl NodeGroup {
	/// Add (or find) a passthrough for an inner input (C++
	/// `add_input_passthrough()`): returns the existing id if the
	/// input is already passed through; otherwise mints a fresh id
	/// (`input.input()`, suffixed `_2`, `_3`, ... until unique, or
	/// `force_id` when given — which must not already exist), adds a
	/// group input with the inner input's type/default/flags, and
	/// records the mapping.
	pub fn add_input_passthrough(&mut self, core: &mut NodeCore, input: InnerInput, force_id: &str) -> String {
		let _ = (core, input, force_id);
		todo!()
	}

	/// Remove the passthrough (and the group input) for an inner
	/// input (C++ `remove_input_passthrough()`); no-op if absent.
	pub fn remove_input_passthrough(&mut self, core: &mut NodeCore, input: &InnerInput) {
		let _ = (core, input);
		todo!()
	}

	/// The inner output node (C++ `get_output_passthrough()`).
	pub fn output_passthrough(&self) -> Option<NodeId> {
		todo!()
	}

	/// Set the inner output node (C++ `set_output_passthrough()`);
	/// the node must belong to the group's context (C++ asserts).
	pub fn set_output_passthrough(&mut self, node: Option<NodeId>) {
		let _ = node;
		todo!()
	}

	/// Whether an inner input is already passed through (C++
	/// `contains_input_passthrough()`).
	pub fn contains_input_passthrough(&self, input: &InnerInput) -> bool {
		let _ = input;
		todo!()
	}

	/// Group input id for an inner input, or empty (C++
	/// `get_id_of_passthrough()`).
	pub fn id_of_passthrough(&self, input: &InnerInput) -> &str {
		let _ = input;
		todo!()
	}

	/// Inner input behind a group input id, if any (C++
	/// `get_input_from_id()`).
	pub fn input_from_id(&self, id: &str) -> Option<&InnerInput> {
		let _ = id;
		todo!()
	}

	/// Fully resolve a group input to the innermost non-group input
	/// (C++ `resolve_input()` / `get_inner()`): loops
	/// [`Self::get_inner`] until the input no longer refers to a
	/// group passthrough.
	pub fn resolve_input(input: InnerInput) -> InnerInput {
		let _ = input;
		todo!()
	}

	/// One resolution step (C++ `get_inner()`): if `input` points at
	/// a group node's passthrough input, rewrite it to the mapped
	/// inner input and return true; false otherwise.
	pub fn get_inner(input: &mut InnerInput) -> bool {
		let _ = input;
		todo!()
	}
}

impl NodeBehavior for NodeGroup {
	/// Human-readable name (C++ `name()`).
	fn name(&self) -> &str {
		"Group"
	}

	/// Stable type id (C++ `id()`).
	fn type_id(&self) -> &str {
		"org.olivevideoeditor.Olive.group"
	}

	/// Categories (C++ `category()` returns `{k_category_unknown}` =
	/// -1, which has no `Category` enum counterpart; modeled as an
	/// empty slice — the group is hidden from the create menu via
	/// the dont-show-in-create-menu flag set in [`create`]).
	fn categories(&self) -> &[Category] {
		&[]
	}

	/// Description (C++ `description()`).
	fn description(&self) -> &str {
		"A group of nodes that is represented as a single node."
	}

	/// Localized input name (C++ `get_input_name()` override):
	/// returns the group's own override name if one was set;
	/// otherwise forwards to the passed-through inner node's input
	/// name (recursively, if that node is itself a group); empty
	/// when the id is not a passthrough.
	///
	/// C++ `retranslate()` additionally recurses `retranslate()`
	/// into every node in the group's context; that traversal is a
	/// graph-level concern and lives with the caller, not here.
	fn input_name<'a>(&self, id: &'a str) -> &'a str {
		let _ = id;
		todo!()
	}

	/// Custom project load (C++ `load_custom()`): reads the
	/// `inputpassthroughs` element (each `inputpassthrough` records
	/// inner node/input/element, passthrough id, custom name, flags,
	/// data type, default value, and a `properties` map of key/value
	/// pairs, deferred into the serialized data's group link list)
	/// and the `outputpassthrough` element (inner node reference,
	/// deferred likewise). Returns true.
	fn load_custom(&mut self, core: &mut NodeCore, reader: &mut dyn crate::serializer::XmlRead) -> bool {
		let _ = (core, reader);
		todo!()
	}

	/// Custom project save (C++ `save_custom()`): writes every
	/// passthrough (inner node reference, input id, element,
	/// passthrough id, display name, flags minus the inner input's
	/// own flags, data type name, default value, and properties) and
	/// the output passthrough node reference.
	fn save_custom(&self, core: &NodeCore, writer: &mut dyn crate::serializer::XmlWrite) {
		let _ = (core, writer);
		todo!()
	}

	/// Post-load fixup (C++ `PostLoadEvent()`): resolves the
	/// deferred group links — re-adds each input passthrough against
	/// the now-live inner nodes and restores its custom flags, name,
	/// data type, default value, and properties — then restores each
	/// group's output passthrough.
	fn post_load(&mut self, core: &mut NodeCore) {
		let _ = core;
		todo!()
	}

	/// Deep copy (C++ `copy()` via `NODE_DEFAULT_FUNCTIONS`);
	/// clones the passthrough table and output reference too.
	fn duplicate(&self, core: &NodeCore) -> Option<Box<dyn NodeBehavior>> {
		let _ = core;
		todo!()
	}
}

/// Constructor (C++ `NodeGroup::NodeGroup()`): no inputs of its own
/// (passthroughs are added dynamically); initializes the output
/// passthrough to unset and sets the dont-show-in-create-menu flag.
pub fn create() -> (NodeCore, Box<dyn NodeBehavior>) {
	todo!()
}

/// Register this node type (C++ factory entry for
/// `org.olivevideoeditor.Olive.group`; note C++ files it under
/// `k_category_unknown`, which is unrepresentable in [`Category`]).
pub fn register(meta: &mut Vec<NodeMeta>) {
	meta.push(NodeMeta {
		type_id: "org.olivevideoeditor.Olive.group",
		name: "Group",
		categories: &[],
		create,
	});
}
